cpp/src/binaryop/compiled/struct_binary_ops.cuh

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "binary_ops.hpp"
+#include "operation.cuh"
+
+#include <cudf/binaryop.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_view.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/table/experimental/row_operators.cuh>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+namespace cudf::binops::compiled::detail {
+template <class T, class... Ts>
+inline constexpr bool is_any_v = std::disjunction<std::is_same<T, Ts>...>::value;
+
+template <class BinaryOperator,
+          typename PhysicalElementComparator =
+            cudf::experimental::row::lexicographic::sorting_physical_element_comparator>
+void apply_struct_binary_op(mutable_column_view& out,
+                            column_view const& lhs,
+                            column_view const& rhs,
+                            bool is_lhs_scalar,
+                            bool is_rhs_scalar,
+                            PhysicalElementComparator comparator = {},
+                            rmm::cuda_stream_view stream         = cudf::default_stream_value)
+{
+  auto compare_orders   = std::vector<order>(lhs.size(),
+                                           is_any_v<BinaryOperator, ops::Greater, ops::GreaterEqual>
+                                               ? order::DESCENDING
+                                               : order::ASCENDING);
+  auto tlhs             = table_view{{lhs}};
+  auto trhs             = table_view{{rhs}};
+  auto table_comparator = cudf::experimental::row::lexicographic::two_table_comparator{
+    tlhs, trhs, compare_orders, {}, stream};
+  auto outd = column_device_view::create(out, stream);
+  auto optional_iter =
+    cudf::detail::make_optional_iterator<bool>(*outd, nullate::DYNAMIC{out.has_nulls()});
+
+  if (is_any_v<BinaryOperator, ops::LessEqual, ops::GreaterEqual>) {
```

Comment:
Use a lambda with an `auto device_comparator` argument so that you can minimize the duplicated code here while remaining "templated" (via `auto`) on the different comparator types. https://godbolt.org/z/d95ddboGW

Reply:
This definitely removed some duplicate code, but net added code and made it slightly less readable. Worth noting, but brobably could have gone either way and not an issue.